igvm/hypervisor.py

```diff
@@ -826,8 +880,50 @@ def device_to_netcat(self, device, size, listener):
         self.run(
             'dd if={0} ibs=1048576 | pv -f -s {1} '
             '| /bin/nc.openbsd -q 1 {2} {3}'
-            .format(device, size, *listener)
+            .format(device, size, listener[0], listener[1])
+        )
+
+    @contextmanager
+    def xfsrestore(self, vm: VM, transaction: Transaction = None):
+        device = self.get_volume_by_vm(vm).path()
+        port = self.netcat_port(device)
+        mount_dir = self.format_vm_storage(vm, transaction)
+
+        # Check if xfsrestore exists, since problem is hidden with
+        # 'nohup', '&' and 'pty=False'
+        self.run('xfsrestore --help', silent=True)
+
+        # xfsrestore args:
+        #  -F: Don't prompt the operator.
+        #  -J: inhibits inventory update
+        # xfsrestore needs to output its logs, otherwise it fails
+        self.run(
+            'nohup /bin/nc.openbsd -l -p {0} '
+            '| xfsrestore -F -J - {1} 2>{1}.log 1>&2 &'
+            .format(port, mount_dir),
+            pty=False,  # Has to be here for background processes
+        )
+        try:
+            yield self.fqdn, port
+        except BaseException:
```

Comment:
Why are you explicelty capturing BaseException here ?

Reply:
This is the same usage as in `netcat_to_device`. I suspect this is done to keep flake8 silent about `cathing everything` error.

Reply:
Let me rephrase the questions: Why do you expect yielding two primitive data types (a string and a integer) to raise a BaseException ?

If it is because of the method `_netcat_port` raising it we should consider using Transaction.

Reply:
From my side, I've just reused the same snippet from https://github.com/innogames/igvm/commit/7ed936d4#diff-d587443efd90ee0cdc812d478981451c4b4a15000f584b1fc06ea3256d154fddR743

Maybe, the contextmanager API is used wrong in the first place, but as far as I know, @TuxPowered42 would like to get rid of Transaction and to move to the context lib.

Reply:
Here's the answer, why `BaseException` is caught https://docs.python.org/3.5/library/contextlib.html#contextlib.contextmanager

> If an unhandled exception occurs in the block, it is reraised inside the generator at the point where the yield occurred.

So, if anything happens at https://github.com/innogames/igvm/pull/176/files#diff-d587443efd90ee0cdc812d478981451c4b4a15000f584b1fc06ea3256d154fddR694, the `@contextmanager` will clean up the netcat server on the target hypervisor.

Reply:
Ah makes sense. Thanks for explaining.